Mathlib/RingTheory/FiniteType.lean

```diff
@@ -706,35 +705,101 @@ end MonoidAlgebra
 
 end MonoidAlgebra
 
+section Orzech
+
+open Submodule in
+/-- **Orzech's theorem** for finitely generated module: if `R` is a commutative ring,
+`M` and `N` are `R`-modules, `M` is finitely generated, `i : N →ₗ[R] M` is injective,
+`f : N →ₗ[R] M` is surjective, then `f` is also injective.
+This is a consequence of Noetherian case
+(`IsNoetherian.injective_of_surjective_of_injective`), which requires that `M` is a
+Noetherian module, but allows `R` to be non-commutative. The proof of this result from
+Noetherian case is adapted from <https://math.stackexchange.com/a/1066110/235999>. -/
```

Comment:
I believe that Stackexchange references are not appropriate, their lifespan is not guaranteed.
Can you please add a standard reference (eg, to Orzech's paper)?

Reply:
Also: While the reference there assumes commutativity to approximate the finitely generated module by a noetherian module, I have the impression that a similar argument would work in the noncommutative case.

Reply:
(There is a paper by Paulo Ribenboim https://eudml.org/doc/111421 that investigates this question. 
Let's say a ring is Orzechian if Orzech's theorem holds for a finitely generated module over that ring. 
Left-noetherian rings are orzechian. 
A theorem of Djokovic asserts that inductive limits of orzechian rings are orzechian, and Ribenboim deduces from this fact that commutative rings are orzechian.)

Reply:
Thank you for the information. Let me list the relevant papers:

1. Orzech, Morris.
*Onto endomorphisms are isomorphisms*.
Amer. Math. Monthly 78 (1971), 357--362.
2. Djoković, D. Ž.
*Epimorphisms of modules which must be isomorphisms*.
Canad. Math. Bull. 16 (1973), 513--515.
3. Ribenboim, Paulo.
*Épimorphismes de modules qui sont nécessairement des isomorphismes*. (French)
Séminaire P. Dubreil, M.-L. Dubreil-Jacotin, L. Lesieur et C. Pisot (24ème année: 1970/71), Algèbre et théorie des nombres, Fasc. 2, Exp. No. 19, 5 pp.
Éditions de l'Académie de la République Socialiste de Roumanie, Bucharest, 1971.

The MathSciNet review of [3] suggests that it only proves that all commutative rings are Orzechian, as in this PR. It's still not known (in these 3 papers) if all non-commutative rings are also Orzechian.

Reply:
> Ribenboim notes the interesting fact: if `R` is orzechian and is a free `R`-module generated by `m` elements and `N` is a a free submodule of `M` which is generated by `n` elements, then `n≤m`.

Isn't it `StrongRankCondition`? So I suspect that Orzechian is in fact equivalent to `StrongRankCondition`?

Reply:
Indeed [IsNoetherianRing.strongRankCondition](https://leanprover-community.github.io/mathlib4_docs/Mathlib/LinearAlgebra/InvariantBasisNumber.html#IsNoetherianRing.strongRankCondition) and [commRing_strongRankCondition](https://leanprover-community.github.io/mathlib4_docs/Mathlib/LinearAlgebra/FreeModule/StrongRankCondition.html#commRing_strongRankCondition) are easily implied by this PR, but I'm not sure whether you can go the other way. I didn't read the proofs in this PR or on Math.SE, but I have the impression that they resemble developments in [this section](https://leanprover-community.github.io/mathlib4_docs/Mathlib/LinearAlgebra/Prod.html#Tunnels-and-tailings). Since you've proven more general results, we can consider removing those.

Reply:
Let me see. Should the new results added in this PR be put into a new file?

Reply:
I don't know whether that counterexample to the strong rank condition is already in mathlib, but it makes sense to add it. (At least, include comments here so that people remember!)

Reply:
The ribenboim reference you mentioned is the one I had indicated you, but the end of the reference probably applies to the Djokovic paper.

Reply:
> I don't know whether that counterexample to the strong rank condition is already in mathlib, but it makes sense to add it. (At least, include comments here so that people remember!)

The examples are mentioned in the answer of <https://math.stackexchange.com/questions/4711904>. I doubt that they are in mathlib, since the module doc of the file containing strong rank condition does not mention the existence of counterexamples.

Reply:
Can you just add this link to that file with a TODO? Maybe somebody will be willing to add the counterexample.

Reply:
> Can you just add this link to that file with a TODO? Maybe somebody will be willing to add the counterexample.

Sure, I'm working on this.

Reply:
I added remarks to corresponding files. Hopefully this suggestion is resolved.